### DISCUSSION TITLE
Is it good idea that rendering a widget consumes it?

### DISCUSSION THREAD

**Discussion by JerzySpendel** [fdehau/tui-rs#523] [General]
I'll copy the trait definition so that we know what we're talking about:

```rust
pub trait Widget {
    fn render(self, area: Rect, buf: &mut Buffer);
}
```

From some other issues I get that you wanted to keep Widgets simple, but why does that have to imply that render method must consume itself? If the trait is defined like that it makes it impossible to have for example vector of different Widget types and call `render` on each of them, because Widget is not object-safe.

So something like this:

```rust
let widgets: Vec<&dyn Widget> = vec![&ChartWidget::new(), &TableWidget::new()];
for widget in widgets {
    widget.render(Rect::default(), some_buffer);
}
```

is not possible, because self's type is not known at compile time. In my opinion it effectively disables polymorphism. Is there any other way I could iterate over widgets and draw them iteratively?

**Comment by vdods** [comment]
Agreed on this point.  It does seem unnecessary.

One could implement Widget for `&YourType` in which case render takes a reference.  But I think (though I don't know and haven't tried), that that doesn't get around the fact that Widget is not an object safe trait.

**Comment by Yevgnen** [comment]
@JerzySpendel Is a way or workaround to make your second snippets work?
